pkg/hs/urbit-king/lib/Urbit/Vere/Ames.hs

```diff
@@ -194,36 +206,86 @@ ames env who isFake enqueueEv stderr = (initialEvents, runAmes)
     drv  <- mkRAcquire start stop
     pure (handleEffect drv mode)
 
-  start :: HasLogFunc e => RIO e AmesDrv
+  start :: RIO e AmesDrv
   start = do
+    mode <- rio (netMode isFake)
+    cachedScryLane <- cache scryLane
+
     aTurfs   <- newTVarIO Nothing
     aDropped <- newTVarIO 0
+    aVersion <- newTVarIO Nothing
+    aVersTid <- trackVersionThread aVersion
     aUdpServ <- udpServ isFake who
-    aRecvTid <- queuePacketsThread aDropped aUdpServ
     aResolvr <- resolvServ aTurfs (usSend aUdpServ) stderr
+    aRecvTid <- queuePacketsThread
+      aDropped
+      aVersion
+      cachedScryLane
+      (send aUdpServ aResolvr mode)
+      aUdpServ
+
     pure (AmesDrv { .. })
 
   hearFailed _ = pure ()
 
-  queuePacketsThread :: HasLogFunc e => TVar Word -> UdpServ -> RIO e (Async ())
-  queuePacketsThread dropCtr UdpServ {..} = async $ forever $ do
-    outcome <- atomically $ do
-      (p, a, b) <- usRecv
-      enqueueEv (EvErr (hearEv p a b) hearFailed)
-    case outcome of
-      Intake -> pure ()
-      Ouster -> do
-        d <- atomically $ do
-          d <- readTVar dropCtr
-          writeTVar dropCtr (d + 1)
-          pure d
-        when (d `rem` packetsDroppedPerComplaint == 0) $
-          logWarn "ames: queue full; dropping inbound packets"
-
-  stop :: AmesDrv -> RIO e ()
+  trackVersionThread :: HasLogFunc e => TVar (Maybe Version) -> RIO e (Async ())
+  trackVersionThread versSlot = async $ forever do
+    scryVersion >>= \case
+      Just v -> do
+        v0 <- readTVarIO versSlot
+        atomically $ writeTVar versSlot (Just v)
+        if (v0 == Just v)
+          then logInfo $ displayShow ("ames: proto version unchanged at", v)
+          else stderr ("ames: protocol version now " <> tshow v)
+
+      Nothing -> logError "ames: could not scry for version"
+
+    threadDelay (10 * 60 * 1_000_000)  -- 10m
+
+  queuePacketsThread :: HasLogFunc e
+                     => TVar Word
+                     -> TVar (Maybe Version)
+                     -> (Ship -> RIO e (Maybe [AmesDest]))
+                     -> (AmesDest -> ByteString -> RIO e ())
+                     -> UdpServ
+                     -> RIO e (Async ())
+  queuePacketsThread dropCtr vers lan forward UdpServ{..} = async $ forever $ do
+      -- port number, host address, bytestring
+    (p, a, b) <- atomically usRecv
+    ver <- readTVarIO vers
+
+    case decode b of
+      Right (pkt@Packet {..}) | ver == Nothing || ver == Just pktVersion -> do
+        logDebug $ displayShow ("ames: bon packet", pkt, showUD $ bytesAtom b)
+
+        if pktRcvr == who
+          then serfsUp p a b
+          else lan pktRcvr >>= \case
+            Just (dest:_) -> forward dest $ encode pkt
```

Comment:
To prevent forwarding loops (#3778), the forwarding logic needs to check if the destination lane is a galaxy (and if we are a galaxy) and exclude our selves. See e59863b in #3764 for an example.